File: python/03_form_fill.py
"""Fill a PDF form with AI vision — works on scanned and non-editable PDFs."""

import os
import sys
import json
import time
import requests

API_KEY = os.environ.get("DEEPREAD_API_KEY")
BASE = "https://api.deepread.tech"

if not API_KEY:
    print("Set DEEPREAD_API_KEY in your environment or .env file")
    sys.exit(1)

file_path = sys.argv[1] if len(sys.argv) > 1 else None
if not file_path:
    print("Usage: python 03_form_fill.py <blank_form.pdf>")
    sys.exit(1)

headers = {"X-API-Key": API_KEY}

# Data to fill into the form
form_fields = json.dumps({
    "full_name": "Jane Smith",
    "date_of_birth": "1990-03-15",
    "address": "123 Main St, San Francisco, CA 94102",
    "phone": "(555) 867-5309",
    "email": "jane.smith@example.com",
    "ssn": "456-78-9012"
})

# Submit form + data
with open(file_path, "rb") as f:
    resp = requests.post(
        f"{BASE}/v1/form-fill",
        headers=headers,
        files={"file": f},
        data={"form_fields": form_fields},
    )
    resp.raise_for_status()

job = resp.json()
job_id = job["id"]
print(f"Submitted: {job_id}")

# Poll with backoff
delay = 3
while True:
    time.sleep(delay)
    result = requests.get(f"{BASE}/v1/form-fill/{job_id}", headers=headers).json()
    status = result["status"]
    print(f"  Status: {status}")

    if status == "completed":
        filled_url = result.get("filled_file_url")
        print(f"\nFilled form: {filled_url}")

        # Download the filled PDF
        if filled_url:
            pdf = requests.get(filled_url).content
            out_path = file_path.replace(".pdf", "_filled.pdf")
            with open(out_path, "wb") as f:
                f.write(pdf)
            print(f"Saved to: {out_path}")
        break
    elif status == "failed":
        print(f"Failed: {result.get('error', 'Unknown error')}")
        break

    delay = min(delay * 1.5, 15)

File: node/03_form_fill.mjs
// Fill a PDF form with AI vision

import { readFileSync, writeFileSync } from "fs";

const API_KEY = process.env.DEEPREAD_API_KEY;
const BASE = "https://api.deepread.tech";
const filePath = process.argv[2];

if (!API_KEY) { console.error("Set DEEPREAD_API_KEY"); process.exit(1); }
if (!filePath) { console.error("Usage: node 03_form_fill.mjs <blank_form.pdf>"); process.exit(1); }

const formFields = JSON.stringify({
  full_name: "Jane Smith",
  date_of_birth: "1990-03-15",
  address: "123 Main St, San Francisco, CA 94102",
  phone: "(555) 867-5309",
  email: "jane.smith@example.com",
});

const form = new FormData();
form.append("file", new Blob([readFileSync(filePath)]), filePath.split("/").pop());
form.append("form_fields", formFields);

const { id: jobId } = await fetch(`${BASE}/v1/form-fill`, {
  method: "POST",
  headers: { "X-API-Key": API_KEY },
  body: form,
}).then(r => r.json());

console.log(`Submitted: ${jobId}`);

let delay = 3000;
while (true) {
  await new Promise(r => setTimeout(r, delay));
  const result = await fetch(`${BASE}/v1/form-fill/${jobId}`, {
    headers: { "X-API-Key": API_KEY },
  }).then(r => r.json());

  console.log(`  Status: ${result.status}`);

  if (result.status === "completed") {
    const filledUrl = result.filled_file_url;
    console.log(`\nFilled form: ${filledUrl}`);
    if (filledUrl) {
      const pdf = await fetch(filledUrl).then(r => r.arrayBuffer());
      const outPath = filePath.replace(".pdf", "_filled.pdf");
      writeFileSync(outPath, Buffer.from(pdf));
      console.log(`Saved to: ${outPath}`);
    }
    break;
  } else if (result.status === "failed") {
    console.error(`Failed: ${result.error}`);
    break;
  }
  delay = Math.min(delay * 1.5, 15000);
}
